### PR TITLE
tests: support different images on nested execution

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This change adds some changes needed to be able to use precreated images on nested tests execution.

Also introduces some minor improvements

### CODE REVIEW

File: tests/lib/nested.sh [hunk 1]
```diff
@@ -239,99 +239,173 @@ cleanup_nested_env(){
...
-        if [ -d "${PWD}/extra-snaps" ]; then
-            while IFS= read -r mysnap; do
-                EXTRA_SNAPS="$EXTRA_SNAPS --snap $mysnap"
-            done <   <(find extra-snaps -name '*.snap')
-        fi
+get_image_path(){
+    echo "$WORK_DIR/image"
+}
+
+get_core_image_name(){
```
Comment (zyga) [inline tests/lib/nested.sh:246]: I'm not sure if having this and the next function really helps.

Comment (sergiocazzolato) [reply tests/lib/nested.sh:246]: I'll update that, thanks for the comment

File: tests/lib/nested.sh [hunk 2]
```diff
@@ -268,99 +268,176 @@ cleanup_nested_env(){
...
+    local IMAGE_URL=$1
+    local IMAGE_NAME=$2
+    local IMAGE_PATH
+    IMAGE_PATH="$(get_image_path)"
+
+    if [[ "$IMAGE_URL" == *.img.xz ]]; then
+        curl -L -o "${IMAGE_PATH}/${IMAGE_NAME}.xz" "$IMAGE_URL"
+        unxz "${IMAGE_PATH}/${IMAGE_NAME}.xz"
+    elif [[ "$IMAGE_URL" == *.img ]]; then
+        curl -L -o "${IMAGE_PATH}/${IMAGE_NAME}" "$IMAGE_URL"
```
Comment (stolowski) [inline tests/lib/nested.sh:329]: Could this be simplified to an unconditional `curl -L -o "${IMAGE_PATH}/${IMAGE_NAME}" "$IMAGE_URL"`, and then just unxz ... if url ends with .xz?

File: tests/lib/nested.sh [hunk 3]
```diff
@@ -268,99 +268,176 @@ cleanup_nested_env(){
...
+    else
+        echo "Image extension not supported, exiting..."
```
Comment (stolowski) [inline tests/lib/nested.sh:331]: Can you print IMAGE_URL as part of this error?

Comment (sergiocazzolato) [reply tests/lib/nested.sh:331]: sure

File: tests/lib/nested.sh (hunk 2 above)
Comment (sergiocazzolato) [reply tests/lib/nested.sh:329]: sure, the only consideration is that unxz just works when the file ends with .xz.
Thanks for the review

**Comment by sergiocazzolato** [comment]
@stolowski thanks for the review, the comments have been addressed.

**Comment by sergiocazzolato** [comment]
@stolowski thanks for the review. I just merged master and triggered the tests.

File: tests/lib/nested.sh [hunk 4]
```diff
@@ -265,111 +265,184 @@ cleanup_nested_env(){
...
-        case "$SPREAD_SYSTEM" in
+get_extra_snaps(){
+    local EXTRA_SNAPS=""
+    local EXTRA_SNAPS_PATH
+    EXTRA_SNAPS_PATH="$(get_extra_snaps_path)"
+
+    if [ -d "$EXTRA_SNAPS_PATH" ]; then
+        while IFS= read -r mysnap; do
+            echo "$mysnap"
+        done <   <(find "$EXTRA_SNAPS_PATH" -name '*.snap')
```
Comment (anonymouse64) [inline tests/lib/nested.sh:308]: unnecessary whitespace, but feel free to apply in a followup
```suggestion
        done < <(find "$EXTRA_SNAPS_PATH" -name '*.snap')
```

File: tests/lib/nested.sh [hunk 5]
```diff
@@ -448,18 +523,24 @@ configure_cloud_init_nested_core_vm_uc20(){
...
+    IMAGE_DIR="$(get_image_dir)"
+    IMAGE_NAME="$(get_image_name core)"
+
+    # As core18 systems use to fail to start the assertion disk when using the
+    # snapshot feature, we copy the original image and use that copy to start
+    # the VM.
+    # Some tests however need to force stop and restart the VM with different
+    # options, so if that env var is set, we will reuse the existing file if it
+    # exists
+    cp "$IMAGE_DIR/$IMAGE_NAME" "$CURRENT_IMAGE"
```
Comment (anonymouse64) [inline tests/lib/nested.sh:543]: why was this moved into the start_nested_core_vm_unit function? the reason I introduced `start_nested_core_vm_unit` at all was so that we could avoid copying the image file and just start the vm systemd unit directly w/o modifying the image file, as there are cloud-init nested spread tests which rely on this functionality. I'm surprised those tests didn't fail when you did this, did you try running the nested cloud-init spread tests with this branch ?

Comment (sergiocazzolato) [reply tests/lib/nested.sh]: Thanks for the clarification, I already fixed it but I had to leave the cp inside the start_nested_core_vm_unit because in that function we calculate the name of the source image to copy. Previouly there was just 1 source image in the system, now and in the following PRs we could have more than 1 in order to reuse them. This is important to speed up in case we dont use kvm and we need to save time when running tests like refresh fundamentals.

Comment (sergiocazzolato) [reply tests/lib/nested.sh]: @anonymouse64 finally I moved back the cp to start_nested_core_vm and also added a comment to explain there the different scenarios for the current image.

**Comment by anonymouse64** [comment]
let's try to remember to squash merge this as there's a lot of git history here now

File: spread.yaml [hunk 6]
```diff
@@ -922,13 +922,11 @@ suites:
...
             BUILD_SNAPD_FROM_CURRENT: '$(HOST: echo "${SPREAD_BUILD_SNAPD_FROM_CURRENT:-false}")'
+            # Download and use an custom image from this url
+            CUSTOM_IMAGE_URL: '$(HOST: echo "${SPREAD_CUSTOM_IMAGE_URL:-}")'
             # Directory where the images and test assets are stored
             WORK_DIR: '/tmp/work-dir'
         manual: true
-        debug: |
-            if [ -f "${WORK_DIR}/serial.log" ]; then
-                cat "${WORK_DIR}/serial.log"
-            fi
```
Comment (anonymouse64) [inline spread.yaml:931]: why delete this ? if we have this here then when preparing a test fails we still get the boot log, which has proven to be useful numerous times

Comment (sergiocazzolato) [reply spread.yaml]: @anonymouse64 Sorry, I though is was needed temporally to see the reboot errors, I already add it again.

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -230,8 +230,7 @@ jobs:
       run: |
           # Register a problem matcher to highlight spread failures
           echo "::add-matcher::.github/spread-problem-matcher.json"
-          export CORE_CHANNEL=edge
```
Comment (anonymouse64) [inline .github/workflows/test.yaml:233]: why this change? what happens when we leave this in?

File: tests/lib/nested.sh [hunk 8]
```diff
@@ -265,111 +265,188 @@ cleanup_nested_env(){
...
+    . "$TESTSLIB"/prepare.sh
+    # shellcheck source=tests/lib/snaps.sh
+    . "$TESTSLIB"/snaps.sh
+
+    local IMAGE_DIR IMAGE_NAME
+    IMAGE_DIR="$(get_image_dir)"
+    IMAGE_NAME="$(get_image_name core)"
+
+    prepare_image_dir
+    if [ ! -f "$IMAGE_DIR/$IMAGE_NAME" ]; then
```
Comment (anonymouse64) [inline tests/lib/nested.sh:364]: if the image name exists then this function just silently exits, correct?

File: .github/workflows/test.yaml (hunk 7 above)
Comment (anonymouse64) [reply .github/workflows/test.yaml:233]: no I did not get an answer about this over IRC, at least I don't remember it if I did get one :smiley:

Comment (sergiocazzolato) [reply .github/workflows/test.yaml:233]: Sorry for the delay in my response, I removed that because it was wrong, the correct var if we need to set the core channel is SPREAD_CORE_CHANNEL, but in this case is ti not needed because by default is edge.

This is the default which is set in the spread.yaml:
CORE_CHANNEL: '$(HOST: echo "${SPREAD_CORE_CHANNEL:-edge}")'

File: tests/lib/nested.sh (hunk 8 above)
Comment (sergiocazzolato) [reply tests/lib/nested.sh:364]: yes, in the following pr, we will check here if the image was previously built, so it can be reused

File: tests/nested/core/hotplug/task.yaml [hunk 9]
```diff
@@ -34,7 +34,9 @@ execute: |
     . "$TESTSLIB/hotplug.sh"
 
     echo "Refresh core snap to desired channel"
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/nested/core/hotplug/task.yaml:36]: sure
